linked_list.c

```diff
@@ -1,105 +1,197 @@
 #include <assert.h>
 #include "linked_list.h"
 
+
 void ll_init(struct ll_node *node) {
     assert(node != NULL);
     // TODO: Implement your solution here.
+
+    node->next = NULL;
+    node->prev = NULL;
 }
 
 bool ll_has_next(struct ll_node *node) {
     assert(node != NULL);
     // TODO: Implement your solution here.
-    return false;
+
+    return node->next != NULL;
 }
 
 bool ll_has_prev(struct ll_node *node) {
     assert(node != NULL);
     // TODO: Implement your solution here.
-    return false;
+
+    return node->prev != NULL;
 }
 
 struct ll_node *ll_next(struct ll_node *node) {
     assert(node != NULL);
     // TODO: Implement your solution here.
+
+    if(ll_has_next(node)) {
+        return node->next;
+    }
     return NULL;
 }
 
 struct ll_node *ll_prev(struct ll_node *node) {
     assert(node != NULL);
     // TODO: Implement your solution here.
+
+    if(ll_has_prev(node)) {
+        return node->prev;
+    }
     return NULL;
 }
 
 size_t ll_size(struct ll_node *head) {
     assert(head != NULL);
     // TODO: Implement your solution here.
-    return 0;
+
+    size_t size = 1;
+    
+    while(ll_has_next(head)) {
+        size++;
+        head = ll_next(head);
+    }
+    return size;
 }
 
 struct ll_node *ll_head(struct ll_node *list) {
     assert(list != NULL);
     // TODO: Implement your solution here.
-    return NULL;
+
+    while(ll_has_prev(list)) {
+        list = ll_prev(list);
+    }
+    return list;
 }
 
 struct ll_node *ll_tail(struct ll_node *list) {
     assert(list != NULL);
     // TODO: Implement your solution here.
-    return NULL;
+
+    while(ll_has_next(list)) {
+        list = ll_next(list);
+    }
+    return list;
 }
 
 struct ll_node *ll_get(struct ll_node *node, size_t index) {
     assert(node != NULL);
     // TODO: Implement your solution here.
-    return NULL;
+
+    // is this assertion needed?
+    // assert(index < size);
```

Comment:
Not really, since the program will likely crash if you try to pass an index off the end of the list anyway.  However, our goal is always to program defensively, so the assertion could be more useful in giving a useful error than a simple segmentation fault.  I left out the explicit assertion in the shell file because scanning the size of the list on every access would make the O(n) `get()` method even more costly than it already is.